python_modules/dagster/dagster/_core/definitions/partition.py

```diff
@@ -972,18 +972,27 @@ def with_partition_key_range(
     def __or__(self, other: "PartitionsSubset") -> "PartitionsSubset[T_str]":
         if self is other:
             return self
+        # Anything | AllPartitionsSubset = AllPartitionsSubset
+        if isinstance(other, AllPartitionsSubset):
+            return cast(PartitionsSubset[T_str], other)  # Is this safe?
```

Comment:
Agree that this is safe, but maybe we can just make the type annotation "PartitionsSubset" instead of "PartitionsSubset[T_str]" to avoid the cast? To be honest I don't know why this class needs to be a Generic -- my suspicion is that this is a holdover from days when we considered letting keys be datetime objects and such.